Add script tracking reserved function names

pfl opened the pull request:
This pull request adds a Coccinelle-Python script that renames functions starting with underscores as specified in #9882. The script takes into account "_current", "_ready_q", "_timeout_q" and "_threads" mentioned in that issue, I'm not sure whether they actually are functions or otherwise just mentioned.

The script hasn't been thoroughly tested yet, and no patches have yet been created either, but they are soon to follow.

himanshujha199640 commented:
Cc @JuliaLawall

JuliaLawall commented:
The rules have a lot of explicit dependencies on other rules.  Actually, they don't need this.  If a python rule has a metavariable from another rule X, then the python rule will only run if the metavariable is bound, and so the python rule intrinsically depends on X.

A minor issue is that it is possible to say "depends on !xxx".  No parentheses are needed.

himanshujha199640 reviewed (changes requested):
Quick glance:
```
himanshu@himanshu-Vostro-3559:~/zephyr$ spatch -D patch --sp-file patrik.cocci --very-quiet  drivers/pinmux/stm32/pinmux_stm32.c
--- drivers/pinmux/stm32/pinmux_stm32.c
+++ /tmp/cocci-output-18172-9236f1-pinmux_stm32.c
@@ -129,7 +129,7 @@ void stm32_setup_pins(const struct pin_c
 	clk = device_get_binding(STM32_CLOCK_CONTROL_NAME);
 
 	for (i = 0; i < pins; i++) {
-		_pinmux_stm32_set(pinconf[i].pin_num,
+		z_pinmux_stm32_set(pinconf[i].pin_num,
 				  pinconf[i].mode,
 				  clk);
 	}
```
What about removing underscore from the function definition ?
```
himanshu@himanshu-Vostro-3559:~/zephyr$ git grep '_pinmux_stm32_set' drivers/pinmux/stm32/pinmux_stm32.c
drivers/pinmux/stm32/pinmux_stm32.c:int _pinmux_stm32_set(u32_t pin, u32_t func,
drivers/pinmux/stm32/pinmux_stm32.c:            _pinmux_stm32_set(pinconf[i].pin_num,

```

pfl commented:
> What about removing underscore from the function definition ?

Yes, it should do that too. Just wondering about the syntax to make this happen right now. Any ideas appreciated.

pfl commented:
> The rules have a lot of explicit dependencies on other rules. Actually, they don't need this. If a python rule has a metavariable from another rule X, then the python rule will only run if the metavariable is bound, and so the python rule intrinsically depends on X.

Yes, I noticed that Coccinelle must be looking at the metavariables after submitting. I'll fix.

> A minor issue is that it is possible to say "depends on !xxx". No parentheses are needed.

True, I'll fix.

JuliaLawall commented:
On Fri, 14 Dec 2018, Patrik Flykt wrote:

>
>       What about removing underscore from the function definition ?
>
> Yes, it should do that too. Just wondering about the syntax to make this
> happen right now. Any ideas appreciated.

You can make a rule that matches a function definition, just like anything
else.

If that wasn't the question, please send a concrete example of what you
would like to do.

julia

>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub, or mute the
> thread.[AAesmkGNfMvOrui97c929edijD4ktohQks5u4-JwgaJpZM4ZSrrI.gif]
>
>
>

pfl commented:
Updated.

himanshujha199640 commented:
> Updated.

Nope! Still the same problem.

So, add this rule too:

```
@r_sign_out@
identifier r_anyfn.fn, r_any_rewrite.z;
type T;
@@
  
    T
  - fn
  + z
    (...) {...}
```

And then you get:
```
--- drivers/pinmux/stm32/pinmux_stm32.c
+++ /tmp/cocci-output-11056-200593-pinmux_stm32.c
@@ -103,7 +103,7 @@ static int stm32_pin_configure(int pin,
  *
  * @return 0 on success, error otherwise
  */
-int _pinmux_stm32_set(u32_t pin, u32_t func, struct device *clk)
+int z_pinmux_stm32_set(u32_t pin, u32_t func, struct device *clk)
 {
 	/* make sure to enable port clock first */
 	if (enable_port(STM32_PORT(pin), clk)) {
@@ -128,7 +128,7 @@ void stm32_setup_pins(const struct pin_c
 	clk = device_get_binding(STM32_CLOCK_CONTROL_NAME);
 
 	for (i = 0; i < pins; i++) {
-		_pinmux_stm32_set(pinconf[i].pin_num,
+		z_pinmux_stm32_set(pinconf[i].pin_num,
 				  pinconf[i].mode,
 				  clk);
```

Please also check for others false positives if any and consider adding `Confidence` level
accordingly. Looks like now it should have no false positives and hence `HIGH`.

Let me know if there are more problems and I know its really difficult and annoying in the
beginning to the frame the script.

Also, I like this approach of RFC for only coccinelle script without spamming other fellow
developers.

Lastly, to make it easier for us, in future attach an example.c file and what you actually want
to do with .c file.

himanshujha199640 commented on scripts/coccinelle/reserved_function_names.cocci:
```diff
@@ -0,0 +1,98 @@
+fn << r_anyfn.fn;
+z;
+@@
+
```
Add here:
```
if fn == '__ASSERT':
          cocci.include_match(False)
```
There is one other variant I found `__ASSERT_NO_MSG`.
Maybe using regex would suffice but can't say, not a python
expert.

Otherwise you get :-/
```
diff -u -p a/pinmux_cc2650.c b/pinmux_cc2650.c
--- a/pinmux_cc2650.c
+++ b/pinmux_cc2650.c
@@ -96,7 +96,7 @@ static int pinmux_cc2650_get(struct devi
 static int pinmux_cc2650_pullup(struct device *dev, u32_t pin,
 				 u8_t func)
 {
-	__ASSERT((func == PINMUX_PULLUP_ENABLE) |
+	z_ASSERT((func == PINMUX_PULLUP_ENABLE) |
 		 (func == PINMUX_PULLUP_DISABLE),
 		 "Pullup mode is invalid");
 
@@ -117,7 +117,7 @@ static int pinmux_cc2650_pullup(struct d
 static int pinmux_cc2650_input(struct device *dev, u32_t pin,
 				u8_t func)
 {
-	__ASSERT((func == PINMUX_INPUT_ENABLED) |
+	z_ASSERT((func == PINMUX_INPUT_ENABLED) |
 		 (func == PINMUX_OUTPUT_ENABLED),
 		 "I/O mode is invalid");
```

ceolin replied:
this should be Z_ASSERT, if you not match these cases you'll not be able to do the change.

pfl replied:
Z_ASSERT added to the list of prefixes in r_any_rewrite.

himanshujha199640 commented on scripts/coccinelle/reserved_function_names.cocci:
```diff
@@ -0,0 +1,98 @@
+
+@r_static_out depends on patch@
+identifier r_staticfn.fn;
+identifier r_static_rewrite.z;
```
`identifier  r_staticfn.fn, r_static_rewrite.z;` is fine.

pfl replied:
Ok, let's go with those once we're experts. Right now my head is spinning less if I have them on separate lines - go figure... :-)

himanshujha199640 commented on scripts/coccinelle/reserved_function_names.cocci:
```diff
@@ -0,0 +1,98 @@
+/// Find function names starting with '__' or '_' and rename them
+// Copyright: (C) 2018 Intel Corporation
+// SPDX-License-Identifier: Apache-2.0
```
// Confidence: High ?

Just check the reports carefully as this script
changes the codebase massivefully.

ceolin replied:
Agree

pfl replied:
Seems to work better now, need to still check the output of this script before setting any confidence level.

pfl replied:
Let's fix the confidence level after another round of updates. There are quite a few _impl_* functions that are somehow missed.

himanshujha199640 commented on scripts/coccinelle/reserved_function_names.cocci:
```diff
@@ -0,0 +1,98 @@
+
+  fn@p (...)
+
+@r_staticfn disable optional_storage@
```
I don't think `disable optional_storage` is useful here.

pfl replied:
Nope, fixed.

ceolin commented:
> > Updated.
> 
> Nope! Still the same problem.
> 
> So, add this rule too:
> 
> ```
> @r_sign_out@
> identifier r_anyfn.fn, r_any_rewrite.z;
> type T;
> @@
>   
>     T
>   - fn
>   + z
>     (...) {...}
> ```
> 
> And then you get:
> 
> ```
> --- drivers/pinmux/stm32/pinmux_stm32.c
> +++ /tmp/cocci-output-11056-200593-pinmux_stm32.c
> @@ -103,7 +103,7 @@ static int stm32_pin_configure(int pin,
>   *
>   * @return 0 on success, error otherwise
>   */
> -int _pinmux_stm32_set(u32_t pin, u32_t func, struct device *clk)
> +int z_pinmux_stm32_set(u32_t pin, u32_t func, struct device *clk)
>  {
>  	/* make sure to enable port clock first */
>  	if (enable_port(STM32_PORT(pin), clk)) {
> @@ -128,7 +128,7 @@ void stm32_setup_pins(const struct pin_c
>  	clk = device_get_binding(STM32_CLOCK_CONTROL_NAME);
>  
>  	for (i = 0; i < pins; i++) {
> -		_pinmux_stm32_set(pinconf[i].pin_num,
> +		z_pinmux_stm32_set(pinconf[i].pin_num,
>  				  pinconf[i].mode,
>  				  clk);
> ```
> 
> Please also check for others false positives if any and consider adding `Confidence` level
> accordingly. Looks like now it should have no false positives and hence `High`.
> 
> Let me know if there are more problems and I know its really difficult and annoying in the
> beginning to the frame the script.
> 
> Also, I like this approach of RFC for only coccinelle script without spamming other fellow
> developers.
> 
> Lastly, to make it easier for us, in future attach an example.c file and what you actually want
> to do with .c file.

The idea is good, but where you want to attach this example ? I don't think this should be part of Zephyr's code base.

himanshujha199640 commented:
> > > Updated.
> > 
> > 
> > Nope! Still the same problem.
> > So, add this rule too:
> > ```
> > @r_sign_out@
> > identifier r_anyfn.fn, r_any_rewrite.z;
> > type T;
> > @@
> >   
> >     T
> >   - fn
> >   + z
> >     (...) {...}
> > ```
> > And then you get:
> > ```
> > --- drivers/pinmux/stm32/pinmux_stm32.c
> > +++ /tmp/cocci-output-11056-200593-pinmux_stm32.c
> > @@ -103,7 +103,7 @@ static int stm32_pin_configure(int pin,
> >   *
> >   * @return 0 on success, error otherwise
> >   */
> > -int _pinmux_stm32_set(u32_t pin, u32_t func, struct device *clk)
> > +int z_pinmux_stm32_set(u32_t pin, u32_t func, struct device *clk)
> >  {
> >  	/* make sure to enable port clock first */
> >  	if (enable_port(STM32_PORT(pin), clk)) {
> > @@ -128,7 +128,7 @@ void stm32_setup_pins(const struct pin_c
> >  	clk = device_get_binding(STM32_CLOCK_CONTROL_NAME);
> >  
> >  	for (i = 0; i < pins; i++) {
> > -		_pinmux_stm32_set(pinconf[i].pin_num,
> > +		z_pinmux_stm32_set(pinconf[i].pin_num,
> >  				  pinconf[i].mode,
> >  				  clk);
> > ```
> > Please also check for others false positives if any and consider adding `Confidence` level
> > accordingly. Looks like now it should have no false positives and hence `High`.
> > Let me know if there are more problems and I know its really difficult and annoying in the
> > beginning to the frame the script.
> > Also, I like this approach of RFC for only coccinelle script without spamming other fellow
> > developers.
> > Lastly, to make it easier for us, in future attach an example.c file and what you actually want
> > to do with .c file.
> 
> The idea is good, but where you want to attach this example ? I don't think this should be part of Zephyr's code base.

No, I meant when you or @pfl have any queries about how to frame any
corner case using coccinelle then provide an example.c file as an attachement
in the email(or wherever possible) with what changes needs to be done to that file.

In this case PR though, I figured it out by examining the reports of the script 
and quickly looking at the referenced Issue about what needs to be done.

Anyway, with the function defintion transformation + other minor bits I proposed,
this script is good to go.

pfl commented:
> @r_sign_out@
> identifier r_anyfn.fn, r_any_rewrite.z;
> type T;
> @@
> T
> - fn
> + z
> (...) {...}

Interesting, so the existing r_any_out wasn't enough to find all occurences. Is there a brief explanation of why this wasn't the case?

himanshujha199640 commented:
> > @r_sign_out@
> > identifier r_anyfn.fn, r_any_rewrite.z;
> > type T;
> > @@
> > T
> > 
> > * fn
> > 
> > 
> > * z
> >   (...) {...}
> 
> Interesting, so the existing r_any_out wasn't enough to find all occurences. Is there a brief explanation of why this wasn't the case?

Quick answer:

` fn@p (...)` doesn't match function definition, instead only `fn (...) {...}` does.
Try the following:
```
@r@
identifier fn =~ "^_.*";
@@

 fn (...) {...}

@script:python@
f << r.fn;
@@

print f
```

pfl commented:
> Quick answer:
> 
> ` fn@p (...)` doesn't match function definition, instead only `fn (...) {...}` does.

Apparently yes. But I can't follow your reply to 100%, here we added another match
`    T
  - fn
  + z
    (...) {...}`
which resulted in functions being renamed as expected. And then I added a variant that depended on r_staticfn, and it seems to have done the trick. I still have to figure out why, though...

pfl commented:
> No, I meant when you or @pfl have any queries about how to frame any
> corner case using coccinelle then provide an example.c file as an attachement
> in the email(or wherever possible) with what changes needs to be done to that file.

I added another commit with a test file and the expected output. This could be wrapped into a Coccinelle script test case or so...

pfl commented:
So far, so good. Added ext/ directory handling into the Python script r_any_rewrite, since the prefixes __arch, __sys, _k, etc. are used by functions in ext/, but the rest of the reserved function names in ext/ should not be mangled. Also fixed up the static function name handling as Coccinelle 1.0.4 does seem to have issues with depends on !(file in "ext"). And besides, it is nowhere documented that the "ext" matches only directory names, we would not like to find out that files or directories called e.g. testexampleext would not be handled due to this.

But anyway, now the next leaky point is that .h files are not updated in include/. All .c files seem to be handled correctly now.

pfl commented:
The remaining problem is that running `coccicheck -m=patch -c=scripts/coccinelle/reserved_function_names.cocci include/ does not produce any patch but only

`Please check for false positives in the output before submitting a patch.
When using "patch" mode, carefully review the patch before submitting it.

coccicheck failed`

Instead, when specifying exactly the .h file in question, like so: `coccicheck -m=patch -c=scripts/coccinelle/reserved_function_names.cocci include/dma.h` gives the expected output of 

`Please check for false positives in the output before submitting a patch.
When using "patch" mode, carefully review the patch before submitting it.

diff -u -p a-project/include/dma.h b-project/include/dma.h
--- a-project/include/dma.h
+++ b-project/include/dma.h
@@ -239,7 +239,7 @@ static inline int dma_reload(struct devi
  */
 __syscall int dma_start(struct device *dev, u32_t channel);
 
-static inline int _impl_dma_start(struct device *dev, u32_t channel)
+static inline int impl_dma_start(struct device *dev, u32_t channel)
 {
 	const struct dma_driver_api *api =
 		(const struct dma_driver_api *)dev->driver_api;
@@ -262,7 +262,7 @@ static inline int _impl_dma_start(struct
  */
 __syscall int dma_stop(struct device *dev, u32_t channel);
 
-static inline int _impl_dma_stop(struct device *dev, u32_t channel)
+static inline int impl_dma_stop(struct device *dev, u32_t channel)
 {
 	const struct dma_driver_api *api =
 		(const struct dma_driver_api *)dev->driver_api;`

What is the correct fix for this?

JuliaLawall commented:
On Tue, 8 Jan 2019, Patrik Flykt wrote:

>
> The remaining problem is that running `coccicheck -m=patch
> -c=scripts/coccinelle/reserved_function_names.cocci include/ does not
> produce any patch but only
>
> `Please check for false positives in the output before submitting a patch.
> When using "patch" mode, carefully review the patch before submitting it.
>
> coccicheck failed`

I'm not able to look into it at the moment, but maybe you can try V=1
(verbose) and DEBUG_FILE="err.log" (place to put extra output).

julia

>
> Instead, when specifying exactly the .h file in question, like so:
> coccicheck -m=patch -c=scripts/coccinelle/reserved_function_names.cocci
> include/dma.h gives the expected output of
>
> `Please check for false positives in the output before submitting a patch.
> When using "patch" mode, carefully review the patch before submitting it.
>
> diff -u -p a-project/include/dma.h b-project/include/dma.h
> --- a-project/include/dma.h
> +++ b-project/include/dma.h
> @@ -239,7 +239,7 @@ static inline int dma_reload(struct devi
> */
> __syscall int dma_start(struct device *dev, u32_t channel);
>
> -static inline int _impl_dma_start(struct device *dev, u32_t channel)
> +static inline int impl_dma_start(struct device *dev, u32_t channel)
> {
> const struct dma_driver_api *api =
> (const struct dma_driver_api *)dev->driver_api;
> @@ -262,7 +262,7 @@ static inline int _impl_dma_start(struct
> */
> __syscall int dma_stop(struct device *dev, u32_t channel);
>
> -static inline int _impl_dma_stop(struct device *dev, u32_t channel)
> +static inline int impl_dma_stop(struct device *dev, u32_t channel)
> {
> const struct dma_driver_api *api =
> (const struct dma_driver_api *)dev->driver_api;`
>
> What is the correct fix for this?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub, or mute the
> thread.[AAesmu9GPRBWlRDb8WeZVtsYjVeWVs45ks5vBO_tgaJpZM4ZSrrI.gif]
>
>
>

pfl commented:
> What is the correct fix for this?

Adding `// Options: --include-headers` to the top of the Coccinelle script, it seems. Also fixed the default renaming to check whether the function name is all upper case and prepend a prefix of 'Z_' in that case.

JuliaLawall commented:
On Tue, 8 Jan 2019, Patrik Flykt wrote:

>
>       What is the correct fix for this?
>
> Adding // Options: --include-headers to the top of the Coccinelle script, it
> seems. Also fixed the default renaming to check whether the function name is
> all upper case and prepend a prefix of 'Z_' in that case.

This is what I suspected.  But I'm surprised that coccicheck would fail
because it found nothing to do, rather than just quietly do nothing.
Perhaps Himanshu can check on this issue.

julia

>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub, or mute the
> thread.[AAesmgsYs1PNHIQK--Ssbd4GyflT-Sc-ks5vBQT6gaJpZM4ZSrrI.gif]
>
>
>

pfl commented:
Then I had to add a new replacement rule r_anyproto_out at the bottom, which will rename function prototypes. With this latest addition we get almost what we want:

```
--- a-project/include/misc/rb.h
+++ b-project/include/misc/rb.h
@@ -74,8 +74,8 @@ typedef void (*rb_visit_t)(struct rbnode
 
 struct rbnode *_rb_child(struct rbnode *node, int side);
 int _rb_is_black(struct rbnode *node);
-void _rb_walk(struct rbnode *node, rb_visit_t visit_fn, void *cookie);
-struct rbnode *_rb_get_minmax(struct rbtree *tree, int side);
+void z_rb_walk(struct rbnode *node, rb_visit_t visit_fn, void *cookie);
+struct rbnode *z_rb_get_minmax(struct rbtree *tree, int side);
```

...but since _rb_child() and _rb_is_black() were not used, our initial rule r_anyfn was not able to detect either of them. More thinking needed...

pfl commented:
> ...but since _rb_child() and _rb_is_black() were not used, our initial rule r_anyfn was not able to detect either of them. More thinking needed...

Running `coccicheck -m=patch -c=scripts/coccinelle/reserved_function_names.cocci . | grep rb_child` gives:
```
-		struct rbnode *ch = _rb_child(node, side);
+		struct rbnode *ch = z_rb_child(node, side);
 struct rbnode *_rb_child(struct rbnode *node, int side);
```
So the function is eventually used elsewhere.

himanshujha199640 commented:
One of those times when coccinelle pisses me off :-/ 

So, debugging step by step:
```
himanshu@himanshu-Vostro-3559:~/zephyr$ cat test.cocci; spatch --sp-file test.cocci drivers/dma/dma_handlers.c
@r@
identifier fn =~ "^_.*";
@@

  fn (...)

@script:python@
f << r.fn;
@@

print f
init_defs_builtins: /usr/local/bin/../lib/coccinelle/standard.h
HANDLING: drivers/dma/dma_handlers.c
```
But when:
```
himanshu@himanshu-Vostro-3559:~/zephyr$ git diff; cat test.cocci; spatch --sp-file test.cocci drivers/dma/dma_handlers.c
diff --git a/drivers/dma/dma_handlers.c b/drivers/dma/dma_handlers.c
index a85dab5726..0cbe58d1cb 100644
--- a/drivers/dma/dma_handlers.c
+++ b/drivers/dma/dma_handlers.c
@@ -11,7 +11,7 @@
  * the validity of the channel ID and returning -errno if it's bogus
  */
 
-Z_SYSCALL_HANDLER(dma_start, dev, channel)
+int bullshit (dma_start, dev, channel)
 {
        Z_OOPS(Z_SYSCALL_DRIVER_DMA(dev, start));
        return _impl_dma_start((struct device *)dev, channel);
@r@
identifier fn =~ "^_.*";
@@

  fn (...)

@script:python@
f << r.fn;
@@

print f
init_defs_builtins: /usr/local/bin/../lib/coccinelle/standard.h
HANDLING: drivers/dma/dma_handlers.c
_impl_dma_start
```

Basiclly coccinelle doesn't recognize `Z_SYSCALL_HANDLER` and no matching occurs.
I would recommend `typedef` appproach to make coccinelle aware but it can't be used here
I think. http://coccinelle.lip6.fr/docs/main_grammar016.html#sec24

@JuliaLawall any workarounds or did I missing something ?

@pfl just use `print` statements for debugging like we use `printk` for the kernel. Will save your
time and always go for `--verbose` output.

And for `--include-headers` stuff, I didn't implement those yet for Zephyr because I couldn't
understand the layout of Zephyr kernel. If anyone is interested then they can lookout for its
implementation linux kernel under `scripts/coccicheck` and do it likewise for Zephyr. Otherwise,
in future when I get some time I would do it myself.

About `coccicheck failed`, please send `spatch --version; ocaml --version`.
You may also try `--debug` option and send its log too.
https://docs.zephyrproject.org/latest/application/coccinelle.html#debugging-coccinelle-smpl-patches

I assume `test_cocci` will not be merged into the Zephyr codebase as it is uninteresting and one can
always just run the script to see the output itself. My suggestion was that whenever you ask for
some suggestion either here or email, just explain us what exactly do you wish to transform by providing
a simple `test.c` and corresponding `test.cocci`.

Hope that helps!

JuliaLawall commented:
On Wed, 9 Jan 2019, Himanshu Jha wrote:

>
> One of those times when coccinelle pisses me off :-/
>
> So, debugging step by step:
>
> himanshu@himanshu-Vostro-3559:~/zephyr$ cat test.cocci; spatch --sp-file test.cocci drivers/dma/dma_handlers.c
> @r@
> identifier fn =~ "^_.*";
> @@
>
>   fn (...)
>
> @script:python@
> f << r.fn;
> @@
>
> print f
> init_defs_builtins: /usr/local/bin/../lib/coccinelle/standard.h
> HANDLING: drivers/dma/dma_handlers.c
>
> But when:
>
> himanshu@himanshu-Vostro-3559:~/zephyr$ git diff; cat test.cocci; spatch --sp-file test.cocci drivers/dma/dma_handlers.c
> diff --git a/drivers/dma/dma_handlers.c b/drivers/dma/dma_handlers.c
> index a85dab5726..0cbe58d1cb 100644
> --- a/drivers/dma/dma_handlers.c
> +++ b/drivers/dma/dma_handlers.c
> @@ -11,7 +11,7 @@
>   * the validity of the channel ID and returning -errno if it's bogus
>   */
>
> -Z_SYSCALL_HANDLER(dma_start, dev, channel)
> +int bullshit (dma_start, dev, channel)
>  {
>         Z_OOPS(Z_SYSCALL_DRIVER_DMA(dev, start));
>         return _impl_dma_start((struct device *)dev, channel);
> @r@
> identifier fn =~ "^_.*";
> @@
>
>   fn (...)
>
> @script:python@
> f << r.fn;
> @@
>
> print f
> init_defs_builtins: /usr/local/bin/../lib/coccinelle/standard.h
> HANDLING: drivers/dma/dma_handlers.c
> _impl_dma_start
>
> Basiclly coccinelle doesn't recognize Z_SYSCALL_HANDLER and no matching occurs.
> I would recommend typedef appproach to make coccinelle aware but it can't be used here
> I think. http://coccinelle.lip6.fr/docs/main_grammar016.html#sec24
>
> @JuliaLawall any workarounds or did I missing something ?

maybe #define Z_SYSCALL_HANDLER int xxx

in mymacros.h and --macro-file-builtins mymacros.h

I'm not sure if it works...

julia

>
> @pfl just use print statements for debugging like we use printk for the kernel. Will save your
> time and always go for --verbose output.
>
> And for --include-headers stuff, I didn't implement those yet for Zephyr because I couldn't
> understand the layout of Zephyr kernel. If anyone is interested then they can lookout for its
> implementation linux kernel under scripts/coccicheck and do it likewise for Zephyr. Otherwise,
> in future when I get some time I would do it myself.
>
> About coccicheck failed, please send spatch --version; ocaml --version.
> You may also try --debug option and send its log too.
> https://docs.zephyrproject.org/latest/application/coccinelle.html#debugging-coccinelle-smpl-patches
>
> I assume test_cocci will not be merged into the Zephyr codebase as it is uninteresting and one can
> always just run the script to see the output itself. My suggestion was that whenever you ask for
> some suggestion either here or email, just explain us what exactly do you wish to transform by providing
> a simple test.c and corresponding test.cocci.
>
> Hope that helps!
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub, or mute the thread.[AAesmlOu1o4XVfJMVM-YhqxEpqd9F-ZIks5vBi-ZgaJpZM4ZSrrI.gif]
>
>
>

pfl commented:
With the lastest changes we get what we want by adding 'T fn@p (...);' to r_anyfn, so the example output is now what we want, i.e.:
```
-struct rbnode *_rb_child(struct rbnode *node, int side);
+struct rbnode *z_rb_child(struct rbnode *node, int side);
-		struct rbnode *ch = _rb_child(node, side);
+		struct rbnode *ch = z_rb_child(node, side);

```
What remains is a way of handling #defines that look just like functions, for example in include/arch/x86/arch.h: `#define _int_latency_start()  do { } while (false)`

pfl commented:
A bit better. Added prefixes to ignore to minimize problems. After applying the patch, one still needs to handle function-looking preprocessor #defines which Coccinelle does not touch, e.g. with `sed 's/__ASSERT/Z_ASSERT/g' include/misc/__assert.h`. After that something is still making a mess with the '_interrupt_stack'.

pfl commented:
> Found the following issues, please fix and resubmit:
> ## identity/email issues
> 
> fa1df1eb1136e04939e00e83484187663765a9eb: author email (Patrik Flykt [patrik.flykt@intel.com](mailto:patrik.flykt@intel.com)) needs to match one of the signed-off-by entries.
> ## commit message syntax issues
> 
> Commit fa1df1eb11:
> 1: UC2 Body does not contain a 'Signed-Off-By' line
> 1: UC3 Title does not follow [subsystem]: [subject] (and should not start with literal subsys:): "tmp"
> 1: UC6 Body has no content, should at least have 1 line.
> 3: B6 Body message is missing

Ignore that, pushed the wrong branch initially. Now fixed.

himanshujha199640 commented:
```
himanshu@himanshu-Vostro-3559:~/zephyr$ cat test.c test.cocci; spatch --sp-file test.cocci test.c

#define _int_latency_start() do { } while (false)

@r@
identifier fn =~ "^_.*";
@@

   #define fn() ...

@script:python@
f << r.fn;
@@

print f
init_defs_builtins: /usr/local/bin/../lib/coccinelle/standard.h
HANDLING: test.c
_int_latency_start
```

I repeat `--include-headers` and all those flags are not supported with current `coccicheck` script.

```
himanshu@himanshu-Vostro-3559:~/zephyr$ ./scripts/coccicheck --mode=patch --verbose=1 --cocci=patrik.cocci  include/arch/x86/arch.h

Please check for false positives in the output before submitting a patch.
When using "patch" mode, carefully review the patch before submitting it.

Processing patrik.cocci
with option(s) " --include-headers"

Message example to submit a patch:
 Find function names starting with '__' or '_' and rename them
 The semantic patch that makes this change is available
 in patrik.cocci.

 More information about semantic patching is available at
 http://coccinelle.lip6.fr/

Running (4 in parallel): /usr/local/bin/spatch -D patch --very-quiet --cocci-file patrik.cocci --include-headers --patch /home/himanshu/zephyr --dir include/arch/x86/arch.h --jobs 4 --chunksize 1
```
You can see how its implemented/used in linux kernel below:
Look for `./arch/*` being included
```
himanshu@himanshu-Vostro-3559:~/linux-next$ make coccicheck V=1 COCCI=scripts/coccinelle/misc/boolinit.cocci M=./drivers/iio/chemical/bme680_core.c 
/bin/bash ./scripts/coccicheck
You have not explicitly specified the mode to use. Using default "report" mode.
Available modes are the following: patch, report, context, org
You can specify the mode with "make coccicheck MODE=<mode>"
Note however that some modes are not implemented by some semantic patches.

Please check for false positives in the output before submitting a patch.
When using "patch" mode, carefully review the patch before submitting it.

Processing boolinit.cocci
with option(s) " --include-headers"

Message example to submit a patch:
 Bool initializations should use true and false.  Bool tests don't need
 comparisons.  Based on contributions from Joe Perches, Rusty Russell
 and Bruce W Allan.

 The semantic patch that makes this report is available
 in scripts/coccinelle/misc/boolinit.cocci.

 More information about semantic patching is available at
 http://coccinelle.lip6.fr/

Running (4 in parallel): /usr/local/bin/spatch -D report --no-show-diff --very-quiet --cocci-file scripts/coccinelle/misc/boolinit.cocci --include-headers --patch . --dir ./drivers/iio/chemical/bme680_core.c -I ./arch/x86/include -I ./arch/x86/include/generated -I ./include -I ./arch/x86/include/uapi -I ./arch/x86/include/generated/uapi -I ./include/uapi -I ./include/generated/uapi --include ./include/linux/kconfig.h --jobs 4 --chunksize 1
```

Now is it implemented can be found in `coccicheck`:
https://github.com/torvalds/linux/blob/master/scripts/coccicheck#L58
https://github.com/torvalds/linux/blob/master/Makefile#L422

With the help of these header directory we get the type and other necessary information but
I couldn't do this zephyr as I wasn't fully confident on the layout of header file organisation.

himanshujha199640 commented:
Wait! I think `--include-headers` can be used but 

```
himanshu@himanshu-Vostro-3559:~/zephyr$ spatch --help | grep 'I'
  -I                             <dir> containing the header files
```
This `-I` is option is not supported yet.

```
himanshu@himanshu-Vostro-3559:~/zephyr$ spatch --help | grep -i 'include'
  --recursive-includes           causes all available include files, both those included in the C file(s) and those included in header files, to be used
  --all-includes                 causes all available include files included in the C file(s) to be used
  --no-includes                  causes not even local include files to be used
  --local-includes               causes local include files to be used
  --include-headers-for-types    use only type information from header files
  --no-include-cache             don't cache parsed include files
  --include-headers              process header files independently
  --include                      file to consider as being included
  --verbose-includes   show on stderr which files are chosen for inclusion
  --relax-include-path            
```

I'm actually confused right now of what flags are benefitical and what are not currently!
Don't no how these flags are interelated to `-I` flag being implemented or not.

JuliaLawall commented:
On Sun, 13 Jan 2019, Himanshu Jha wrote:

>
>       maybe #define Z_SYSCALL_HANDLER int xxx in mymacros.h and
>       --macro-file-builtins mymacros.h I'm not sure if it works...
>
> No. We can't.
>
> kernel/include/syscall_handler.h:#define Z_SYSCALL_HANDLER(...) \
> kernel/include/syscall_handler.h-       _SYSCALL_CONCAT(__SYSCALL_HANDLER, \
> kernel/include/syscall_handler.h-                       _SYSCALL_NARG(__VA_A
> RGS__))(__VA_ARGS__)

I'm not sure to understand.  The definition of the macro doesn't matter.

julia

himanshujha199640 commented:
On Sun, Jan 13, 2019 at 07:36:11AM -0800, JuliaLawall wrote:
> 
> 
> On Sun, 13 Jan 2019, Himanshu Jha wrote:
> 
> >
> >       maybe #define Z_SYSCALL_HANDLER int xxx in mymacros.h and
> >       --macro-file-builtins mymacros.h I'm not sure if it works...
> >
> > No. We can't.
> >
> > kernel/include/syscall_handler.h:#define Z_SYSCALL_HANDLER(...) \
> > kernel/include/syscall_handler.h-       _SYSCALL_CONCAT(__SYSCALL_HANDLER, \
> > kernel/include/syscall_handler.h-                       _SYSCALL_NARG(__VA_A
> > RGS__))(__VA_ARGS__)
> 
> I'm not sure to understand.  The definition of the macro doesn't matter.

OK. But it didn't work as you expected.

himanshu@himanshu-Vostro-3559:~/zephyr$ cat test.cocci; spatch --sp-file test.cocci drivers/dma/dma_handlers.c --macro-file kernel/include/syscall_handler.h
@r@
identifier fn =~ "^_.*";
@@

   fn (...)

@script:python@
f << r.fn;
@@

print f
init_defs_builtins: /usr/local/bin/../lib/coccinelle/standard.h
init_defs: kernel/include/syscall_handler.h
HANDLING: drivers/dma/dma_handlers.c


-- 
Himanshu Jha
Undergraduate Student
Department of Electronics & Communication
Guru Tegh Bahadur Institute of Technology

himanshujha199640 commented:
> The end of standard.h has: #define SYSCALL_DEFINE1(func, t1, a1) \ asmlinkage unsigned long func(t1 a1) #define SYSCALL_DEFINE2(func, t1, a1, t2, a2) \ asmlinkage unsigned long func(t1 a1, t2 a2) #define SYSCALL_DEFINE3(func, t1, a1, t2, a2, t3, a3) \ asmlinkage unsigned long func(t1 a1, t2 a2, t3 a3) ... Maybe you can do the same thing.

OK. I can do that but there are just too many of these in Zephyr.
Another problem is that `Z_SYSCALL_HANDLER` is a variadic macro.
https://docs.zephyrproject.org/1.12.0/kernel/usermode/syscalls.html

Otherwise let me know the declaration of marco and I will send out the
patch.

JuliaLawall commented:
On Wed, 16 Jan 2019, Himanshu Jha wrote:

>
>       The end of standard.h has: #define SYSCALL_DEFINE1(func, t1, a1)
>       \ asmlinkage unsigned long func(t1 a1) #define
>       SYSCALL_DEFINE2(func, t1, a1, t2, a2) \ asmlinkage unsigned long
>       func(t1 a1, t2 a2) #define SYSCALL_DEFINE3(func, t1, a1, t2, a2,
>       t3, a3) \ asmlinkage unsigned long func(t1 a1, t2 a2, t3 a3) ...
>       Maybe you can do the same thing.
>
> OK. I can do that but there are just too many of these in Zephyr.
> Another problem is that Z_SYSCALL_HANDLER is a variadic macro.
> https://docs.zephyrproject.org/1.12.0/kernel/usermode/syscalls.html
>
> Otherwise let me know the declaration of marco and I will send out the
> patch.

Actually, I tried it, and it seemed to be quite fine with:

#define Z_SYSCALL_HANDLER int xxx

and then  -macro-file mac.h

julia

pfl commented:
The 'report' mode now depends on the function renaming, so it will (should) detect the same amount of changes. The script now properly transforms one or two underscores, leaving the ones starting with three to be manually inspected. Also, the function renaming got a bit more work, a missing underscore is added if needed and upper case function (or macro) names is better taken into account.

A new experimental "mode" was added to the script giving a list of name transformations. It might come in handy if only the functions in a particular subdirectory need to be updated, but this again relies on a (shell) script or similar to mop up the changed names in other directories. The issues is how to apply all the changes in sane ways to the rest of the code...

himanshujha199640 commented:
I tried the script but it doesn't handle every aspect as discussed in previous comments.

Let's just focus on `include/arch/x86/arch.h` for now and experiment on this only. This
has all sorts of variants we need to catch.

Few things I would like to state:

* Add print before begining of the scripts to identify what cases were matched and what not.
* There are types that coccinelle doesn't understand `ALWAYS_INLINE` so use `--macro-file` approach.
* Lot of `#define` variants needs to be handled using disjucntions.
* changes mode doesn't look necessary since you can always use report mode. But whatever ...
* Since you made separate rule for catching static, should we put `disable optional_storage` on first rule then ?
 
If you find something weird in transformation which doesn't conform the SmPL grammar, please
report on coccinelle mailing list. There are issues with coccinelle sometimes due to parsing/transformation which developers report and it gets fixed within few days.

pfl commented:
> I tried the script but it doesn't handle every aspect as discussed in previous comments.

I'm about to handle the ones in above comments now with the previous issues fixed.

> Let's just focus on `include/arch/x86/arch.h` for now and experiment on this only. This
> has all sorts of variants we need to catch.
> 
> Few things I would like to state:
> 
>     * Add print before begining of the scripts to identify what cases were matched and what not.
>     * There are types that coccinelle doesn't understand `ALWAYS_INLINE` so use `--macro-file` approach.

I could not quite follow your discussion with Julia on the subject, and I don't really understand how to write  the ALWAYS_INLINE to be undestood by Coccinelle. Can you provide an example of what needs to be done here?
 
>     * Lot of `#define` variants needs to be handled using disjucntions.

Again an example would be nice, I'm not sure even where this is to be implemented. Last time I understood that #define MACRO_XXX wasn't detected at all by Coccinelle.

>     * changes mode doesn't look necessary since you can always use report mode. But whatever ...

Currently it's a hack to just produce a list of  things that are transformed. If it can be figured out in time how a foolproof script can be created, it will be tossed away.

>     * Since you made separate rule for catching static, should we put `disable optional_storage` on first rule then ?

I'm not sure, I need to verify that the functions get detected that way also.

> If you find something weird in transformation which doesn't conform the SmPL grammar, please
> report on coccinelle mailing list. There are issues with coccinelle sometimes due to parsing/transformation which developers report and it gets fixed within few days.

Sure thing, the problem is though that at least I'm not a grammar or SmPL expert at all, and I really have a hard time figuring out whether Coccinelle has a bug, the grammar has a bug or whether my Coccinelle syntax is just bogus...

Really the main problem right now is how to create a set of changes that are 100% complete and won't fail the compilation or linking step. It can be done by a perfect Coccinelle script, or then in parts so that e.g. kernel/ directory and associated header files are changed, but only those all over the source code tree. If done in parts, it should be easier to verify, as there are less things changed.

himanshujha199640 commented:
On Wed, Jan 30, 2019 at 05:16:50PM +0000, Patrik Flykt wrote:
> > I tried the script but it doesn't handle every aspect as discussed in previous comments.
> 
> I'm about to handle the ones in above comments now with the previous issues fixed.
> 
> > Let's just focus on `include/arch/x86/arch.h` for now and experiment on this only. This
> > has all sorts of variants we need to catch.
> > 
> > Few things I would like to state:
> > 
> >     * Add print before begining of the scripts to identify what cases were matched and what not.
> >     * There are types that coccinelle doesn't understand `ALWAYS_INLINE` so use `--macro-file` approach.
> 
> I could not quite follow your discussion with Julia on the subject, and I don't really understand how to write  the ALWAYS_INLINE to be undestood by Coccinelle. Can you provide an example of what needs to be done here?

himanshu@himanshu-Vostro-3559:~/zephyr$ cat mymacros.h; spatch --sp-file test.cocci drivers/dma/dma_handlers.c --macro-file mymacros.h
#define Z_SYSCALL_HANDLER int xxx
init_defs_builtins: /usr/local/bin/../lib/coccinelle/standard.h
init_defs: mymacros.h
HANDLING: drivers/dma/dma_handlers.c
_impl_dma_start
_impl_dma_stop

Here, I created a file mymacros.h which contains type info for
`Z_SYSCALL_HANDLER` which wasn't parsed by coccinelle. And
we pass this header file with `--macro-file` so coccinelle can
extract the type info from it.

Likewise, if `ALWAYS_INLINE` is not detected(please check) then
do something similar for it.

> >     * Lot of `#define` variants needs to be handled using disjucntions.
> 
> Again an example would be nice, I'm not sure even where this is to be implemented. Last time I understood that #define MACRO_XXX wasn't detected at all by Coccinelle.

I meant there are:

* #define _int_latency_start()  do { } while (false)
* #define _VECTOR_ARG(irq_p)      _IRQ_CONTROLLER_VECTOR_MAPPING(irq_p)
* #define _NANO_ERR_SPURIOUS_INT           (0)
  ...

so these need to matched and transformed accordingly.

> >     * changes mode doesn't look necessary since you can always use report mode. But whatever ...
> 
> Currently it's a hack to just produce a list of  things that are transformed. If it can be figured out in time how a foolproof script can be created, it will be tossed away.

OK.

> >     * Since you made separate rule for catching static, should we put `disable optional_storage` on first rule then ?
> 
> I'm not sure, I need to verify that the functions get detected that way also.

It is just for safe-play IMO.
First rule shouldn't match any static stuff, and only second
should.

> > If you find something weird in transformation which doesn't conform the SmPL grammar, please
> > report on coccinelle mailing list. There are issues with coccinelle sometimes due to parsing/transformation which developers report and it gets fixed within few days.
> 
> Sure thing, the problem is though that at least I'm not a grammar or SmPL expert at all, and I really have a hard time figuring out whether Coccinelle has a bug, the grammar has a bug or whether my Coccinelle syntax is just bogus...

Neither am I.

I just refer SmPL grammar when in trouble or look at existing
examples in coccinelle repo or kernel repo.

There is a bug only if something doesn't behave as described
in the SmPL. And if you feel something stupid is done by coccinelle
then report on coccinelle mailing list with test.cocci + test.c

You have picked a corner where everytime I test your script
something new pops up. I imagine how much more complex would
it get to catch every corner.


-- 
Himanshu Jha
Undergraduate Student
Department of Electronics & Communication
Guru Tegh Bahadur Institute of Technology

pfl commented:
With the lastest we still have a remaining problem with Coccinelle. When running `coccicheck -m=patch -c=scripts/coccinelle/reserved_function_names.cocci  kernel/sem.c` on kernel/sem.c, parts of the modifications are found, but not any of the `_impl`* functions:

`--- a-project/kernel/sem.c
+++ b-project/kernel/sem.c
@@ -58,20 +58,20 @@ SYS_INIT(init_sem_module, PRE_KERNEL_1,
 void _impl_k_sem_init(struct k_sem *sem, unsigned int initial_count,
 		      unsigned int limit)
 {
-	__ASSERT(limit != 0U, "limit cannot be zero");
-	__ASSERT(initial_count <= limit, "count cannot be greater than limit");
+	Z_ASSERT(limit != 0U, "limit cannot be zero");
+	Z_ASSERT(initial_count <= limit, "count cannot be greater than limit");
`
@himanshujha199640 what is going wrong here?

pfl commented:
> As explained earlier coccinelle doesn't recognise `Z_SYSCALL_HANDLER`.

Ah, missed that it was hiding behind Z_SYSCALL_HANDLER! I had that file created, but of course not this time around.

pfl commented:
b9c5628 is a shell script for going through all macros and creating the needed .h file.
91cbac2 was done as an example should someone decide to apply changes for a subset of the directories, this script applies all the changes to the whole directory tree

himanshujha199640 commented:
> [b9c5628](https://github.com/zephyrproject-rtos/zephyr/commit/b9c56283d53f257ea281504fbb39a5ffd1d726b4) is a shell script for going through all macros and creating the needed .h file. It also includes the macro file for all functions at the time this patch was created.

Why are you collecting **all** the macro definitions and passing through `--macro-file-buitins` ?
Seems like bogus to me as they are parsed and handled without any problems.
If it were for only `Z_SYSCALL_*` (and others which are not handled by spatch) then it was
fine.

To check which maros are not supported, look here: https://github.com/coccinelle/coccinelle/blob/master/standard.h

Also, I assume these 50k lines of collected macros are not going into the codebase, right ?

In the end, the change should not break kernel build, and whether you do it using coccinelle combined
with any scripting language. But if you want something to be put in `scripts/coccinelle/` then it should
not be bogus and easy to use by developers or bots.

JuliaLawall commented:
On Thu, 7 Feb 2019, Himanshu Jha wrote:

>
>       b9c5628 is a shell script for going through all macros and
>       creating the needed .h file. It also includes the macro file for
>       all functions at the time this patch was created.
>
> Why are you collecting all the macro definitions and passing through
> --macro-file-buitins ?

This doesn't sound like a good idea.  Code that is macro expanded can't be
transformed.  You only want the macro definitions that are needed.  YOu
can often get a good approximation using

spatch --parse-c dirname

You may need to unlimit the stacksize before doing this.  Iterating this a
few times, updating the macro file in between according to what is
complained about, should be enough to get a high coverage rate.

julia

> Seems like bogus to me as they are parsed and handled without any problems.
> If it were for only Z_SYSCALL_* (and others which are not handled by spatch)
> then it was
> fine.
>
> To check which maros are not supported, look here:
> https://github.com/coccinelle/coccinelle/blob/master/standard.h
>
> Also, I assume these 50k lines of collected macros are not going into the
> codebase, right ?
>
> In the end, the change should not break kernel build, and whether you do it
> using coccinelle combined
> with any scripting language. But if you want something to be put in
> scripts/coccinelle/ then it should
> not be bogus and easy to use by developers or bots.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub, or mute the
> thread.[AAesmvWlROmsWpCzOSm11RMujUp0ud5jks5vLAHxgaJpZM4ZSrrI.gif]
>
>
>

himanshujha199640 commented:
> > Why are you collecting **all** the macro definitions and passing through `--macro-file-buitins` ?
> > Seems like bogus to me as they are parsed and handled without any problems.
> > If it were for only `Z_SYSCALL_*` (and others which are not handled by spatch) then it was
> > fine.
> 
> At least the Z_SYSCALL* ones are not detected by Coccinelle, but without going through them all manually, I can't tell which Coccinelle detects and which not. Thus the brute-force approach of collecting them all, which I don't want to do, but there is quite a lot code. I'll probably drop the included macro file, one for only kernel/ is quite a bit smaller. Any ideas on how to do this automatically?

In such a case, I would use `spatch --parse-c <dir/filename` as suggested by @JuliaLawall 
That's an amazing method you may try and catch the bad ones.

I believe rather than lingering on making "perfect" script, we should target small and work
on individual subsystems sequentially.

pfl commented:
Noticed that extern functions were not found, fixed that in the first rule. Also the macro file now contains only the macros discovered with brute force in kernel. So far so good.

But when I do 

`/usr/bin/spatch -D patch --very-quiet --cocci-file scripts/coccinelle/reserved_function_names.cocci --include-headers --dir /home/pflykt/projects/zephyr --macro-file scripts/coccinelle/coccinelle_macros.h kernel/ | grep  abort_thread_timeout`

I get two different outputs for this and a few more functions:

```
-static inline int _abort_thread_timeout(struct k_thread *thread)
+static inline int abort_thread_timeout(struct k_thread *thread)
-			(void)_abort_thread_timeout(thread);
+			(void)z_abort_thread_timeout(thread);
```
The first change is correct in kernel/nclude/timeout_q.h, but the second one in kernel/thread.c. Does this now mean that Coccinelle forgets that the function is static in between, but why would it do that? What might the fix be for this behavior?

JuliaLawall commented:
On Fri, 8 Feb 2019, Patrik Flykt wrote:

>
> Noticed that extern functions were not found, fixed that in the first rule.
> Also the macro file now contains only the macros discovered with brute force
> in kernel. So far so good.
>
> But when I do
>
> /usr/bin/spatch -D patch --very-quiet --cocci-file
> scripts/coccinelle/reserved_function_names.cocci --include-headers --dir
> /home/pflykt/projects/zephyr --macro-file
> scripts/coccinelle/coccinelle_macros.h kernel/ | grep abort_thread_timeout
>
> I get two different outputs for this and a few more functions:
>
> -static inline int _abort_thread_timeout(struct k_thread *thread)
> +static inline int abort_thread_timeout(struct k_thread *thread)
> -			(void)_abort_thread_timeout(thread);
> +			(void)z_abort_thread_timeout(thread);
>
> The first change is correct in kernel/nclude/timeout_q.h, but the second one
> in kernel/thread.c. Does this now mean that Coccinelle forgets that the
> function is static in between, but why would it do that? What might the fix
> be for this behavior?

Are you expecting a change in a header file to be taken into account in
processing a .c file that includes the header file?  That won't happen.  I
guess it could happen if you use --in-place, but there is no guarantee
that the header file will be treated before the .c file.  Maybe you need
--all-includes to cause all directly included headers to be treated with
the .c file that includes them?  Or --recursive-includes to get headers
that are included by other headers.  This will cause a major performance
degradation due to treating a lot more code.

julia


>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub, or mute the
> thread.[AAesmv-cgYP9g4mixe2hIeC_IcyEYRHYks5vLdqPgaJpZM4ZSrrI.gif]
>
>
>

pfl commented:
> Or --recursive-includes to get headers that are included by other headers. This will cause a major performance degradation due to treating a lot more code.

Adding a --recursive-includes seems to do the job!

pfl committed:
scripts/coccinelle: Add a mode to list possible changes

Add a mode to only list possible changes to the reserved_function_names.cocci
script. This may be useful if only a subset of the functions, let's say only
in a specific directory, need to be rewritten. One can then take the output
of this new mode and feed it to a suitable sed or shell script. The output
is also useful for script substitution verification purposes.

Signed-off-by: Patrik Flykt <patrik.flykt@intel.com>

pfl commented:
With the lastest modifications this starts to work as expected. There is one issue remaining that doesn't matter, which is that in patch mode for example all _abort_thread_timeout() are changed to abort_thread_timeout(). But if the same is run in report mode, the same command will suggest some of _abort_thread_timeout() to be changed to abort_thread_timeout() as expected, but also sometimes the static nature of the function is missing, and z_abort_thread_timeout() is proposed instead. I have no idea why this is happening.

The exact command for producing this is:
```
spatch -D patch --very-quiet \
     --cocci-file $ZEPHYR_BASE/scripts/coccinelle/reserved_function_names.cocci \
     --include-headers     --recursive-includes \
     --dir $ZEPHYR_BASE \
     --macro-file scripts/coccinelle/coccinelle_macros.h kernel \
 | grep abort_thread_timeout
```

where `-D patch` is substituted with `-D report`.

pfl commented:
Added another script to force exported function names in kernel/include follow non-static naming even though the functions are declared static.

pfl committed:
scripts/coccinelle: Helper script to apply subset of function changes

If one needs to update only a part of the reserved function names,
the result needs to be applied to the rest of the source code tree.
This is a script that takes as input a list of changed names and
applies them to all files. The input to the script can be created
byt the additional 'changes' mode for the Coccinelle script
reserved_function_names.cocci like so:

spatch -D changes \
--cocci-file $ZEPHYR_BASE/scripts/coccinelle/reserved_function_names.cocci \
--include-headers \
--recursive-includes \
--dir $ZEPHYR_BASE \
--macro-file scripts/coccinelle/coccinelle_macros.h \
kernel/ | sort | uniq > kernel.changes

Apply the changes from kernel/ to the whole source code tree with:

scripts/coccinelle/update_changes.sh kernel.changes

Signed-off-by: Patrik Flykt <patrik.flykt@intel.com>

pfl committed:
scripts/coccinelle: Add script tracking reserved function names

Add a script that tracks reserved function names that start with
one or two underscore chracters. There exists a few cases that
are prepended with three underscores, but these need to be
investigatged and tranformed manually.

When the function is static, remove any leading underscores. In
all other cases look up the function prefixes from a dictionary
and substitute according to the dictionary value or 'z_' if the
prefix is not in the dictionary. Since ext/ is not to be updated
but uses functions defined by the rest of the source code, add
an extra condition in the Python script r_any_rewrite for the
default prefix 'z_' in order not to make such updates this
directory. The same applies for static functions; here a
!(file in "ext") could have been used, but since that part is not
functioning properly with Coccinelle 1.0.4, the same Pyhon script
check is added there.

Add a 'report' mode to the script that can be used by developer/CI
and a 'patch' mode that should do the heavy lifting. As the report
mode has the same dependencies as the patch mode, the same number
of issues will be detected.

Run spatch with --recursive-includes and --macro-file to catch all
locations that need to be renamed.

Signed-off-by: Patrik Flykt <patrik.flykt@intel.com>

nashif commented:
stale, closing. Reopen when there are updates...
